src/librustc_mir/hair/cx/mod.rs

```diff
@@ -105,11 +105,7 @@ impl<'a,'tcx:'a> Cx<'a, 'tcx> {
 
     pub fn needs_drop(&mut self, ty: Ty<'tcx>, span: Span) -> bool {
         if self.infcx.type_moves_by_default(ty, span) {
-            // FIXME(#21859) we should do an add'l check here to determine if
-            // any dtor will execute, but the relevant fn
-            // (`type_needs_drop`) is currently factored into
-            // `librustc_trans`, so we can't easily do so.
-            true
+            self.tcx.type_needs_drop(ty)
```

Comment:
Note, that here infer context is available, and thus it should possible to monomorphise the type (see/copy `rustc::middle::infer::InferCtxt::expr_ty_adjusted`).